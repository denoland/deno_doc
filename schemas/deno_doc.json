{
  "$id": "https://deno.land/schemas/deno_doc.json",
  "$schema": "http://json-schema.org/draft-07/schema",
  "description": "A JSON representation of a deno_doc documentation structure",
  "type": "array",
  "items": {
    "$ref": "#/definitions/DocNode"
  },
  "definitions": {
    "DocNode": {
      "type": "object",
      "required": [
        "kind",
        "name",
        "location"
      ],
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "function",
            "variable",
            "enum",
            "class",
            "typeAlias",
            "namespace",
            "interface",
            "import"
          ]
        },
        "name": {
          "type": "string"
        },
        "location": {
          "$ref": "#/definitions/Location"
        },
        "jsDoc": {
          "type": "string"
        }
      },
      "anyOf": [
        {
          "required": [
            "functionDef"
          ],
          "properties": {
            "kind": {
              "const": "function"
            },
            "functionDef": {
              "$ref": "#/definitions/FunctionDef"
            }
          }
        },
        {
          "required": [
            "variableDef"
          ],
          "properties": {
            "kind": {
              "const": "variable"
            },
            "variableDef": {
              "$ref": "#/definitions/VariableDef"
            }
          }
        },
        {
          "required": [
            "enumDef"
          ],
          "properties": {
            "kind": {
              "const": "enum"
            },
            "enumDef": {
              "$ref": "#/definitions/EnumDef"
            }
          }
        },
        {
          "required": [
            "classDef"
          ],
          "properties": {
            "kind": {
              "const": "class"
            }
          },
          "classDef": {
            "$ref": "#/definitions/ClassDef"
          }
        },
        {
          "required": [
            "typeAliasDef"
          ],
          "properties": {
            "kind": {
              "const": "typeAlias"
            }
          },
          "typeAliasDef": {
            "$ref": "#/definitions/TypeAliasDef"
          }
        },
        {
          "required": [
            "namespaceDef"
          ],
          "properties": {
            "kind": {
              "const": "namespace"
            }
          },
          "namespaceDef": {
            "$ref": "#/definitions/NamespaceDef"
          }
        },
        {
          "required": [
            "interfaceDef"
          ],
          "properties": {
            "kind": {
              "const": "interface"
            }
          },
          "interfaceDef": {
            "$ref": "#/definitions/interfaceDef"
          }
        },
        {
          "required": [
            "importDef"
          ],
          "properties": {
            "kind": {
              "const": "import"
            },
            "importDef": {
              "$ref": "#/definitions/ImportDef"
            }
          }
        }
      ]
    },
    "Location": {
      "type": "object",
      "required": [
        "filename",
        "line",
        "col"
      ],
      "properties": {
        "filename": {
          "type": "string"
        },
        "line": {
          "type": "number"
        },
        "col": {
          "type": "number"
        }
      }
    }
  }
}
