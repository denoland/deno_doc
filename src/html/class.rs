use super::util::*;
use crate::html::parameters::render_params;
use std::fmt::Write;

pub fn render_class(doc_node: &crate::DocNode, ctx: &RenderContext) -> String {
  let class_def = doc_node.class_def.as_ref().unwrap();

  let ctx = &RenderContext {
    current_type_params: class_def
      .type_params
      .iter()
      .map(|def| def.name.clone())
      .collect::<std::collections::HashSet<String>>(),
    ..ctx.clone()
  };

  // TODO: class items

  format!(
    r#"<div class="doc_block_items">{}{}{}{}</div>"#,
    super::jsdoc::render_docs(&doc_node.js_doc, true, false, ctx),
    render_constructors(&class_def.constructors, &doc_node.name, ctx),
    super::types::render_type_params(&class_def.type_params, ctx),
    render_index_signatures(&class_def.index_signatures, ctx),
  )
}

fn render_constructors(
  constructors: &[crate::class::ClassConstructorDef],
  name: &str,
  ctx: &RenderContext,
) -> String {
  if constructors.is_empty() {
    return String::new();
  }

  let items = constructors
    .iter()
    .enumerate()
    .map(|(i, constructor)| {
      let id = name_to_id("constructor", &i.to_string());

      // TODO: tags, render constructor params
      doc_entry(&id, name, "()", constructor.js_doc.doc.as_deref(), ctx)
    })
    .collect::<String>();

  section("Constructors", &items)
}

fn render_index_signatures(
  index_signatures: &[crate::class::ClassIndexSignatureDef],
  ctx: &RenderContext,
) -> String {
  if index_signatures.is_empty() {
    return String::new();
  }

  let items = index_signatures.iter().enumerate().fold(
    String::new(),
    |mut output, (i, index_signature)| {
      let id = name_to_id("index_signature", &i.to_string());

      let readonly = index_signature
        .readonly
        .then_some("<span>readonly </span>")
        .unwrap_or_default();

      let ts_type = index_signature
        .ts_type
        .as_ref()
        .map(|ts_type| {
          format!(": {}", super::types::render_type_def(ts_type, ctx))
        })
        .unwrap_or_default();

      write!(
        output,
        r#"<div class="doc_item" id="{id}">{}{readonly}[{}]{ts_type}</div>"#,
        anchor(&id),
        render_params(&index_signature.params, ctx),
      )
      .unwrap();
      output
    },
  );

  section("Index Signatures", &items)
}
