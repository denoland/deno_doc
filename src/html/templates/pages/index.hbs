{{~> pages/html_head html_head_ctx ~}}
{{~> category_panel categories_panel ~}}
<div>
  {{~> pages/top_nav ~}}

  <div id="content">
    <main>
      {{~#if usage~}}
        {{~> usages_large usage ~}}
      {{~/if~}}

      {{~#if module_doc~}}
        {{~> module_doc module_doc ~}}
      {{~/if~}}

      {{~#if overview~}}
        {{~> symbol_content overview ~}}
      {{~/if~}}
    </main>

    {{~> toc toc_ctx ~}}

    {{~> pages/search_results ~}}
  </div>
</div>
</div>
</body>
</html>
