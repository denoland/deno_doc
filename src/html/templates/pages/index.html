{{ call pages/html_head.html with html_head_ctx }}

{{ call index_sidepanel.html with sidepanel_ctx }}

<div id="content" class="with_sidepanel">
  {{ call breadcrumbs.html with breadcrumbs_ctx }}
  {{ call pages/search_bar.html with search_ctx }}

  <main>
    {#- TODO(bartlomieju): call into template here #}

    {{ if module_doc }}
      {{ call module_doc.html with module_doc }}
    {{ endif }}
  </main>

  {{ call pages/search_results.html with search_ctx }}
<div>
</body>
</html>
