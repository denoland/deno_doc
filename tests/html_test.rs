// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.

use deno_ast::ModuleSpecifier;
use deno_doc::html::pages::SymbolPage;
use deno_doc::html::*;
use deno_doc::DocNode;
use deno_doc::DocParser;
use deno_doc::DocParserOptions;
use deno_graph::source::LoadFuture;
use deno_graph::source::LoadOptions;
use deno_graph::source::LoadResponse;
use deno_graph::source::Loader;
use deno_graph::BuildOptions;
use deno_graph::CapturingModuleAnalyzer;
use deno_graph::GraphKind;
use deno_graph::ModuleGraph;
use futures::future;
use indexmap::IndexMap;
use std::fs;
use std::rc::Rc;

struct SourceFileLoader {}

impl Loader for SourceFileLoader {
  fn load(
    &self,
    specifier: &ModuleSpecifier,
    _options: LoadOptions,
  ) -> LoadFuture {
    let result = if specifier.scheme() == "file" {
      let path = specifier.to_file_path().unwrap();
      fs::read(path)
        .map(|content| {
          Some(LoadResponse::Module {
            specifier: specifier.clone(),
            maybe_headers: None,
            content: content.into(),
          })
        })
        .map_err(|err| err.into())
    } else {
      Ok(None)
    };
    Box::pin(future::ready(result))
  }
}

struct EmptyResolver {}

impl HrefResolver for EmptyResolver {
  fn resolve_path(
    &self,
    current: UrlResolveKind,
    target: UrlResolveKind,
  ) -> String {
    href_path_resolve(current, target)
  }

  fn resolve_global_symbol(&self, _symbol: &[String]) -> Option<String> {
    None
  }

  fn resolve_import_href(
    &self,
    _symbol: &[String],
    _src: &str,
  ) -> Option<String> {
    None
  }

  fn resolve_usage(&self, current_resolve: UrlResolveKind) -> Option<String> {
    current_resolve
      .get_file()
      .map(|current_file| current_file.path.to_string())
  }

  fn resolve_source(&self, _location: &deno_doc::Location) -> Option<String> {
    None
  }
}

async fn get_files(subpath: &str) -> IndexMap<ModuleSpecifier, Vec<DocNode>> {
  let files = fs::read_dir(
    std::env::current_dir()
      .unwrap()
      .join("tests")
      .join("testdata")
      .join(subpath),
  )
  .unwrap();

  let source_files: Vec<ModuleSpecifier> = files
    .into_iter()
    .map(|entry| {
      let entry = entry.unwrap();
      ModuleSpecifier::from_file_path(entry.path()).unwrap()
    })
    .collect();

  let loader = SourceFileLoader {};
  let analyzer = CapturingModuleAnalyzer::default();
  let mut graph = ModuleGraph::new(GraphKind::TypesOnly);
  graph
    .build(
      source_files.clone(),
      &loader,
      BuildOptions {
        module_analyzer: &analyzer,
        ..Default::default()
      },
    )
    .await;

  let parser = DocParser::new(
    &graph,
    &analyzer,
    DocParserOptions {
      diagnostics: false,
      private: false,
    },
  )
  .unwrap();

  let mut source_files = source_files.clone();
  source_files.sort();
  let mut doc_nodes_by_url = IndexMap::with_capacity(source_files.len());
  for source_file in source_files {
    let nodes = parser.parse_with_reexports(&source_file).unwrap();
    doc_nodes_by_url.insert(source_file, nodes);
  }

  doc_nodes_by_url
}

#[tokio::test]
async fn html_doc_files() {
  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: None,
      composable_output: false,
    },
    get_files("single").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css",
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn html_doc_files_rewrite() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");
  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    get_files("multiple").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/A.html",
      "./~/A.prototype.html",
      "./~/B.html",
      "./~/B.prototype.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Baz.foo.html",
      "./~/Baz.html",
      "./~/Foo.bar.html",
      "./~/Foo.html",
      "./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html",
      "./~/Foo.prototype.foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "./~/Hello.html",
      "./~/Hello.world.html",
      "./~/index.html",
      "foo/~/index.html",
      "foo/~/x.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css"
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.bar.html").unwrap());
    insta::assert_snapshot!(files
      .get("./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html")
      .unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.world.html").unwrap());
    insta::assert_snapshot!(files.get("./~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/x.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn symbol_group() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let mut files = vec![];

  {
    for (short_path, doc_nodes) in &ctx.doc_nodes {
      let symbol_pages =
        generate_symbol_pages_for_module(&ctx, short_path, doc_nodes);

      files.extend(symbol_pages.into_iter().map(
        |symbol_page| match symbol_page {
          SymbolPage::Symbol {
            breadcrumbs_ctx,
            symbol_group_ctx,
            toc_ctx,
          } => {
            let root = ctx.href_resolver.resolve_path(
              UrlResolveKind::Symbol {
                file: short_path,
                symbol: &symbol_group_ctx.name,
              },
              UrlResolveKind::Root,
            );

            let html_head_ctx = pages::HtmlHeadCtx::new(
              &root,
              &symbol_group_ctx.name,
              ctx.package_name.as_ref(),
              Some(short_path),
            );

            Some(pages::SymbolPageCtx {
              html_head_ctx,
              symbol_group_ctx,
              breadcrumbs_ctx,
              toc_ctx,
            })
          }
          SymbolPage::Redirect { .. } => None,
        },
      ));
    }
  }

  let mut files_json = serde_json::to_string_pretty(&files).unwrap();
  files_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let symbol_group_json_path = testdata_path.join("symbol_group-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let symbol_group_json_path =
    testdata_path.join("symbol_group-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let symbol_group_json_path = testdata_path.join("symbol_group.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_group_json_path, &files_json).unwrap();
  }

  let symbol_group_json = fs::read_to_string(symbol_group_json_path).unwrap();

  assert_eq!(files_json, symbol_group_json);
}

#[tokio::test]
async fn symbol_search() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let search_index = generate_search_index(&ctx);
  let mut file_json = serde_json::to_string_pretty(&search_index).unwrap();
  file_json.push('\n');

  let symbol_search_json_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("symbol_search.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_search_json_path, &file_json).unwrap();
  }

  let symbol_search_json = fs::read_to_string(symbol_search_json_path).unwrap();

  assert_eq!(file_json, symbol_search_json);
}

#[tokio::test]
async fn module_doc() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    FileMode::Single,
    doc_nodes_by_url,
  )
  .unwrap();

  let mut module_docs = vec![];

  for (short_path, doc_nodes) in &ctx.doc_nodes {
    let render_ctx =
      RenderContext::new(&ctx, doc_nodes, UrlResolveKind::File(short_path));
    let module_doc = jsdoc::ModuleDocCtx::new(&render_ctx, short_path);

    module_docs.push(module_doc);
  }

  let mut file_json = serde_json::to_string_pretty(&module_docs).unwrap();
  file_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let module_docs_json_path = testdata_path.join("module_doc-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let module_docs_json_path = testdata_path.join("module_doc-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let module_docs_json_path = testdata_path.join("module_doc.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&module_docs_json_path, &file_json).unwrap();
  }

  let module_docs_json = fs::read_to_string(module_docs_json_path).unwrap();
  assert_eq!(file_json, module_docs_json);
}
