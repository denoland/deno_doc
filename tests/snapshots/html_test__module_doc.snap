---
source: tests/html_test.rs
expression: module_docs
---
[
  {
    "deprecated": null,
    "sections": {
      "id": "module_doc",
      "docs": "<div class=\"markdown\"><p>Some docs</p>\n<div class=\"alert alert-note\"><div><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"14\" height=\"14\" viewBox=\"0 0 24 24\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2\" stroke-linecap=\"round\" stroke-linejoin=\"round\">\n  <path stroke=\"none\" d=\"M0 0h24v24H0z\" fill=\"none\"></path>\n  <path d=\"M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0\"></path>\n  <path d=\"M12 9h.01\"></path>\n  <path d=\"M11 12h1v4h1\"></path>\n</svg>\nNote</div><div><p>\nUseful information that users should know, even when skimming content.</p>\n</div></div>\n<div class=\"alert alert-tip\"><div><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"14\" height=\"14\" viewBox=\"0 0 24 24\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2\" stroke-linecap=\"round\" stroke-linejoin=\"round\">\n  <path stroke=\"none\" d=\"M0 0h24v24H0z\" fill=\"none\"></path>\n  <path d=\"M3 12h1m8 -9v1m8 8h1m-15.4 -6.4l.7 .7m12.1 -.7l-.7 .7\"></path>\n  <path d=\"M9 16a5 5 0 1 1 6 0a3.5 3.5 0 0 0 -1 3a2 2 0 0 1 -4 0a3.5 3.5 0 0 0 -1 -3\"></path>\n  <path d=\"M9.7 17l4.6 0\"></path>\n</svg>\nTip</div><div><p>\nHelpful advice for doing things better or more easily.</p>\n</div></div>\n<div class=\"alert alert-important\"><div><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"14\" height=\"14\" viewBox=\"0 0 24 24\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2\" stroke-linecap=\"round\" stroke-linejoin=\"round\">\n  <path stroke=\"none\" d=\"M0 0h24v24H0z\" fill=\"none\"></path>\n  <path d=\"M18 4a3 3 0 0 1 3 3v8a3 3 0 0 1 -3 3h-5l-5 3v-3h-2a3 3 0 0 1 -3 -3v-8a3 3 0 0 1 3 -3h12z\"></path>\n  <path d=\"M12 8v3\"></path>\n  <path d=\"M12 14v.01\"></path>\n</svg>\nImportant</div><div><p>\nKey information users need to know to achieve their goal.</p>\n</div></div>\n<div class=\"alert alert-warning\"><div><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"14\" height=\"14\" viewBox=\"0 0 24 24\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2\" stroke-linecap=\"round\" stroke-linejoin=\"round\">\n  <path stroke=\"none\" d=\"M0 0h24v24H0z\" fill=\"none\"></path>\n  <path d=\"M12 9v4\"></path>\n  <path d=\"M10.363 3.591l-8.106 13.534a1.914 1.914 0 0 0 1.636 2.871h16.214a1.914 1.914 0 0 0 1.636 -2.87l-8.106 -13.536a1.914 1.914 0 0 0 -3.274 0z\"></path>\n  <path d=\"M12 16h.01\"></path>\n</svg>\nWarning</div><div><p>\nUrgent info that needs immediate user attention to avoid problems.</p>\n</div></div>\n<div class=\"alert alert-caution\"><div><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"14\" height=\"14\" viewBox=\"0 0 24 24\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2\" stroke-linecap=\"round\" stroke-linejoin=\"round\">\n  <path stroke=\"none\" d=\"M0 0h24v24H0z\" fill=\"none\"></path>\n  <path d=\"M12.802 2.165l5.575 2.389c.48 .206 .863 .589 1.07 1.07l2.388 5.574c.22 .512 .22 1.092 0 1.604l-2.389 5.575c-.206 .48 -.589 .863 -1.07 1.07l-5.574 2.388c-.512 .22 -1.092 .22 -1.604 0l-5.575 -2.389a2.036 2.036 0 0 1 -1.07 -1.07l-2.388 -5.574a2.036 2.036 0 0 1 0 -1.604l2.389 -5.575c.206 -.48 .589 -.863 1.07 -1.07l5.574 -2.388a2.036 2.036 0 0 1 1.604 0z\"></path>\n  <path d=\"M12 8v4\"></path>\n  <path d=\"M12 16h.01\"></path>\n</svg>\nCaution</div><div><p>\nAdvises about risks or negative outcomes of certain actions.</p>\n</div></div>\n</div>",
      "sections": []
    }
  },
  {
    "deprecated": null,
    "sections": {
      "id": "module_doc",
      "docs": null,
      "sections": [
        {
          "header": {
            "title": "Functions",
            "anchor": {
              "id": "Functions"
            },
            "href": null,
            "doc": null
          },
          "content": {
            "kind": "namespace_section",
            "content": [
              {
                "id": "namespace_x",
                "anchor": {
                  "id": "namespace_x"
                },
                "tags": [],
                "doc_node_kind_ctx": [
                  {
                    "kind": "Function",
                    "char": "f",
                    "title": "Function",
                    "title_lowercase": "function",
                    "title_plural": "Functions"
                  }
                ],
                "href": ".././foo/~/x.html",
                "name": "x",
                "docs": null,
                "deprecated": false
              }
            ]
          }
        },
        {
          "header": {
            "title": "Variables",
            "anchor": {
              "id": "Variables"
            },
            "href": null,
            "doc": null
          },
          "content": {
            "kind": "namespace_section",
            "content": [
              {
                "id": "namespace_default",
                "anchor": {
                  "id": "namespace_default"
                },
                "tags": [],
                "doc_node_kind_ctx": [
                  {
                    "kind": "Variable",
                    "char": "v",
                    "title": "Variable",
                    "title_lowercase": "variable",
                    "title_plural": "Variables"
                  }
                ],
                "href": ".././foo/~/default.html",
                "name": "default",
                "docs": "<div class=\"markdown_summary\"><p>The default export item.</p>\n</div>",
                "deprecated": false
              }
            ]
          }
        }
      ]
    }
  }
]
